src/coreclr/jit/lower.cpp

```diff
@@ -3280,6 +3279,57 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
     return nullptr;
 }
 
+//----------------------------------------------------------------------------------------------
+// LowerSelect: Lower a GT_SELECT node.
+//
+// Arguments:
+//     select - The node
+//
+// Return Value:
+//     The next node to lower.
+//
+GenTree* Lowering::LowerSelect(GenTreeConditional* select)
+{
+    GenTree* cond     = select->gtCond;
+    GenTree* trueVal  = select->gtOp1;
+    GenTree* falseVal = select->gtOp2;
+
+    // Replace SELECT cond 1/0 0/1 with (perhaps reversed) cond
+    if (cond->OperIsCompare() && ((trueVal->IsIntegralConst(0) && falseVal->IsIntegralConst(1)) ||
+                                  (trueVal->IsIntegralConst(1) && falseVal->IsIntegralConst(0))))
+    {
+        assert(select->TypeIs(TYP_INT, TYP_LONG));
+
+        LIR::Use use;
+        if (BlockRange().TryGetUse(select, &use))
+        {
+            if (trueVal->IsIntegralConst(0))
+            {
+                GenTree* reversed = comp->gtReverseCond(cond);
+                assert(reversed == cond);
+            }
+
+            GenTree* newNode = cond;
+            if (select->TypeIs(TYP_LONG))
+            {
+                GenTree* cast = comp->gtNewCastNode(TYP_LONG, newNode, false, TYP_LONG);
+                BlockRange().InsertAfter(newNode, cast);
```

Comment:
Why cast here is needed? if we know that we return either 0 or 1 - can we leave it TYP_INT (sort of legal after lowering)

Reply:
Hmm, indeed it looks like compare nodes support being `TYP_LONG` typed, so I can just retype it. Will fix

Reply:
[New diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=165602&view=ms.vss-build-web.run-extensions-tab), a bit better now, but I guess it was rare that we needed this cast